### DISCUSSION TITLE
Does pyinstaller do dead code elimination/code optimizaiton?

### DISCUSSION THREAD

**Discussion by baseplate-admin** [pyinstaller/pyinstaller#8948] [PyInstaller] [answered]
Hey, Thanks for the work on `pyinstaller`

In the JavaScript/C++ world there's the concept of tree-shaking (dead code elimination), I was wondering if `pyinstaller` offers something of this nature? (or maybe is willing to add in the roadmap?)

**Comment by rokm** [comment, marked as answer]
Not really. 

The import analysis (based on `modulegraph`) might help us avoid collecting unused modules of a package, although in practice, due to indirect imports (e.g., programmatic imports with dynamically constructed module names) we often need to resort to collecting all modules from a package just to have things work. But there's no pruning of let's say unused functions or classes within the modules, etc. (Plus the contents of binary extensions and shared libraries cannot be split/pruned anyway).

> or maybe is willing to add in the roadmap?

As far as I'm concerned, the answer is no. Just keeping things running around here takes enough effort. 

From my perspective, it would make more sense to prune your code (perhaps whole virtual environment?) using third-party tree-shaking tool (if it exists for python), *test it unfrozen*, and only freeze it with PyInstaller once you are reasonably sure that pruning did not break anything.

**Reply by baseplate-admin** [reply]
> The import analysis (based on modulegraph) might help us avoid collecting unused modules of a package, although in practice, due to indirect imports (e.g., programmatic imports with dynamically constructed module names) we often need to resort to collecting all modules from a package just to have things work.

It makes sense, it would be a lot of trouble on it's own.


> As far as I'm concerned, the answer is no. Just keeping things running around here takes enough effort.

Thanks for the work you guys are doing around here. `pyinstaller` really is an awesome tool.


> (if it exists for python)

This is the problem, because there is no tool in python that does this.


----

all in all thank you for your input on this matter :) and merry Christmas

**Reply by bwoodsend** [reply]
If such a tool were to be created, I suspect that it would be more effective if used some kind of `sys.settrace()` based runtime profiling instead of the mostly static approach we use for imports. Either way sounds like a complete nightmare to me though.
